### PR TITLE
Policy sync RFC

### PR DESCRIPTION
**Pull request by sergioifg94**
[View document](https://github.com/sergioifg94/architecture/blob/policy-sync-rfc/rfcs/0007-policy-sync-v1.md)

### CODE REVIEW

**Review by Boomatang — COMMENTED** [review]
I have a few comments, from them the main takeaway can be asked as, is there a strong argument that I am missing for wanting to do this work?

File: rfcs/0004-policy-sync-v1.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+* Policies targeting the Multicluster Gateway, defined in the hub cluster and
+  reconciled by the Multicluster Gateway Controller
+* Policies targeting the downstream Gateway, defined in the spoke clusters and
+  reconciled by the downstream Gateway controllers.
+
+In a realistic multicluster scenario where multiple spoke clusters are present, the management of these policies can become tedious and error-prone, as policies have
+to be defined in the hub cluster, as well as replicated in the multiple spoke clusters.
+
+As Kuadrant users:
+* Gateway-admin has a set of homogeneous clusters and needs to apply per cluster rate limits across the entire set.
```
[the remark this answers is not in the document]
Comment (sergioifg94) [reply rfcs/0004-policy-sync-v1.md:38]: From the point of view of a user who wants to create and manage policies for their Gateway, the clusters are homogeneous. The underlying infrastructure is irrelevant for that purpose.

As for the case of applying specific policies to a subset of the clusters, that's why the hierarchy and overriding section is for. A user can create a RateLimitPolicy in the hub cluster to apply in all clusters, and create a RateLimitPolicy in only one of the clusters that overrides the synced Policy.

Basically, the purpose of this RFC is to allow to create policies that target all spoke clusters, in a centralised way, with the possibility of them being overriden

File: rfcs/0004-policy-sync-v1.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+of conflicts, override Policies that contain this annotation.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Third party Policy support
+
+In order for a Policy to be supported for syncing, the MGC must have permissions
+to watch/list/get the resource, and the implementation of the downstream Gateway
+controller must be aware of the `policy-synced` annotation.
```
[the remark this answers is not in the document]
Comment (sergioifg94) [reply rfcs/0004-policy-sync-v1.md:124]: I'm not familiar with that concept, I'll give it a look and see how it affects this RFC

File: rfcs/0004-policy-sync-v1.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+# Rationale and alternatives
```
[the remark this answers is not in the document]
Comment (sergioifg94) [reply rfcs/0004-policy-sync-v1.md:126]: There are many ways to sync resources across clusters, but the goal of this RFC is not only to sync policies, but to manage the synced resources from the hub cluster alongside the policies that target the Multicluster Gateway. 

These alternatives you listed would leave it in the hands of the user to manage the syncing of the policies, or add extra dependencies to the controller, whereas the solution proposed in this RFC would rely on OCM which is already the main framework used to place resources in the spoke clusters.

File: rfcs/0004-policy-sync-v1.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+The ability for the Multicluster Gateway Controller to sync policies defined in
+the hub cluster downstream to the spoke clusters, therefore allowing all policies
+to be defined in the same place. These policies will be reconciled by the downstream
+Gateway controller.
+
+# Nomenclature
+
+* Policy: When refering to a Policy, this document is refering to a Gateway API
+  policy as defined in the Policy Attachment Model. The Multicluster Gateway Controller
+  relies on [OCM]() as a Multicluster solution, which defines its own unrelated
```
[the remark this answers is not in the document]
Comment (sergioifg94) [reply rfcs/0004-policy-sync-v1.md:20]: Good catch, I'll add it to the PR

File: rfcs/0004-policy-sync-v1.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Boomatang) [reply rfcs/0004-policy-sync-v1.md:38]: At some point the capacity of the underlying infrastructure becomes relevant, and I believe this point is at the level of the gateway or gateway class. What the actual underlying infrastructure is (AWS, GCP), that is irrelevant for this purpose.

As of today the RateLimitPolicy and AuthPolicy don't support defaults and overrides, which I am assuming would be required for this. There is talks of getting that support there, so I am ok with this RFC being a thing before that is there. 

Where I do see a problem with the use case of using defaults and overrides, the Kuadrant Policies can only target one resource. And I have not heard any talks about changing this. With the example of creating a RateLimitPolicy on the hub targeting the gateway and is copied to the spokes. If the Gateway owner wanted to override that policy on a per spoke bases, the overrides would need to be added to the RateLimitingPolicy attached to the HTTPRoute. Which maybe owned by different orgs. 

To create the overrides in a RateLimitPolicy on the Gateway, it would require a RateLimitPolicy targeting the GatewayClass which is not supported today.

[the remark this answers is not in the document]
Comment (sergioifg94) [reply rfcs/0004-policy-sync-v1.md:38]: I think there's a mixup about the concept of overrides. When I talk about overrides in this RFC I mean adding support in the Kuadrant operator to ignore a policy that has been synced from the hub if it conflicts with a policy created in the spoke. That support would be part of the implementation of this RFC

Maybe I could add to the document an item to the nomenclature section to avoid confusion

File: rfcs/0004-policy-sync-v1.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sergioifg94) [reply rfcs/0004-policy-sync-v1.md:126]: I'll add these options to the list of alternatives to make sure it's clear why OCM is the best alternative for our situation

File: rfcs/0004-policy-sync-v1.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (maleck13) [reply rfcs/0004-policy-sync-v1.md:38]: > Where I do see a problem with the use case of using defaults and overrides, the Kuadrant Policies can only target one resource. And I have not heard any talks about changing this. With the example of creating a RateLimitPolicy on the hub targeting the gateway and is copied to the spokes. If the Gateway owner wanted to override that policy on a per spoke bases, the overrides would need to be added to the RateLimitingPolicy attached to the HTTPRoute. Which maybe owned by different orgs.

If they wanted to do something per cluster, they would go back to creating the policy on a per cluster basis. This is what they have to do now. 
If they want something to be replicated accross many clusters with the same definition but define that targeting the multicluster gateway, they would define the policy in the hub targeting the multicluster gateway

[the remark this answers is not in the document]
Comment (maleck13) [reply rfcs/0004-policy-sync-v1.md:38]: Some of this really is also to do with the fact we do not let the user set a counter strategy in anyway. 
IE ```counterStrategy``` for limit A is clusterUnique for limit B is shared. being able to define the counter strategy would allow you decide if a somethign should be shared or unique
Also it is possible in RLP to set a limit that only get triggered "when" a certain value is passed. So if you had a unique value being passed from each gateway you could use that to distinguish limits (IE if gateway A set limit to x every y minutes)
Finally another option here is for the policy sync to add some form of label to identify the cluster it came from, these label could be injected into the config for the WASMPlugin and used in conditions.

All of this is to say, that the syncing of the policy is not necessarilly where we want to or where we need to solve changing a set of limits for a particular cluster

[the remark this answers is not in the document]
Comment (maleck13) [reply rfcs/0004-policy-sync-v1.md:38]: https://github.com/Kuadrant/kuadrant-operator/issues/271

**Comment by sergioifg94** [comment]
@Boomatang @maleck13 Pushed a commit addressing the comments

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1 @@
+.idea
```
Comment (pehala) [inline .gitignore:1]: Do we need that in the architecture repo and if so, do we need this in this PR?

File: rfcs/0004-policy-sync-v1.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+The support for resources that the controller can sync is limited by the following:
+* The controller ServiceAccount must have permission to watch, list, and get the
+  resource to be synced
+* The resource must implement the Policy schema:
+    * Have a `.spec.targetRef` field
+
+When a Policy is configured to be synced in a GatewayClass, the Multicluster
+Gateway Controller starts watching events on the resources, and propagates changes
+by placing the policy in the spoke clusters, with the following mutations:
```
Comment (pehala) [inline rfcs/0004-policy-sync-v1.md:70]: Will there be any annotations added to the original policy as well? I think it would be useful if I could trace the spoke policies from some annotation, similar to how Gateway `kuadrant.io/gateway-clusters` works

File: rfcs/0004-policy-sync-v1.md [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Process overview
+
+#### Dynamic Policy watches
+
+The Multicluster Gateway Controller reconciles parameters referenced by the
+GatewayClass of a Gateway. A new field is added to the parameters that allows
```
Comment (pehala) [inline rfcs/0004-policy-sync-v1.md:82]: Will I have to add that I want to sync AP/RLP to every gateway class I create?

File: rfcs/0004-policy-sync-v1.md [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+### Policy Hierarchy 
+
+In order to avoid conflict with Policies created directly in the spoke clusters,
+a hierarchy must be defined to prioritise those Policies.
+
+The controller will set the `kuadrant.io/policy-synced` annotation on the policy
+when placing it in the spoke cluster.
+
+The Kuadrant operator will be aware of the presence of this annotation, and, in case
+of conflicts, override Policies that contain this annotation.
```
Comment (pehala) [inline rfcs/0004-policy-sync-v1.md:117]: Will this be reported back to the original policy? I worry about traceability of the override

File: .gitignore (hunk 5 above)
Comment (sergioifg94) [reply .gitignore:1]: We don't need it, I think that might have been automatically included when editing it with an IntellJ editor. I'll remove it

File: rfcs/0004-policy-sync-v1.md (hunk 6 above)
Comment (sergioifg94) [reply rfcs/0004-policy-sync-v1.md:70]: That's a good idea. Would it be an annotation to reference the downstream policies or simply to inform that it has been synced?

File: rfcs/0004-policy-sync-v1.md (hunk 7 above)
Comment (sergioifg94) [reply rfcs/0004-policy-sync-v1.md:82]: Not necessarily, as that is configured in a ConfigMap, it would be possible to have multiple gateway classes referencing the single ConfigMap so the configuration doesn't have to be repeated

File: rfcs/0004-policy-sync-v1.md (hunk 8 above)
Comment (sergioifg94) [reply rfcs/0004-policy-sync-v1.md:117]: That's a good idea, it could be reported with the `Overriden` status condition proposed in the RFC (https://github.com/Kuadrant/architecture/blob/main/rfcs/0004-policy-status.md#conditions) I'll add it to the document

File: rfcs/0004-policy-sync-v1.md (hunk 6 above)
Comment (pehala) [reply rfcs/0004-policy-sync-v1.md:70]: I would say the former, a reference would be nice

Comment (sergioifg94) [reply rfcs/0004-policy-sync-v1.md:70]: Yes, I agree. I'll add it to the document

**Comment by sergioifg94** [comment]
Pushed some changes with @pehala's suggestions
